File: sea-shell-service-spring-web-flux/sea-shell-rest-service/src/main/java/org/jesperancinha/shell/webflux/service/fork/SeaShellLowerRecursiveTask.java
package org.jesperancinha.shell.webflux.service.fork;

import lombok.AllArgsConstructor;
import lombok.Builder;
import org.jesperancinha.shell.webflux.data.SeaShellCostumeDto;
import org.jesperancinha.shell.webflux.repo.ShellLowerRepository;
import org.jesperancinha.shell.webflux.service.SeaShellConverter;

import java.util.concurrent.RecursiveTask;

@Builder
@AllArgsConstructor
public class SeaShellLowerRecursiveTask extends RecursiveTask<SeaShellCostumeDto> {

    private final ShellLowerRepository lowerRepository;
    private final SeaShellCostumeDto costumeDto;

    @Override
    protected SeaShellCostumeDto compute() {
        costumeDto.setLowerDto(SeaShellConverter.toLowerDto(lowerRepository.findLowerByIdBlock(costumeDto.getLowerId())));
        return costumeDto;
    }
}

File: sea-shell-service-spring-web-flux/src/main/kotlin/org/jesperancinha/shell/webflux/service/fork/SeaShellLowerRecursiveTask.kt
package org.jesperancinha.shell.webflux.service.fork

import lombok.AllArgsConstructor
import lombok.Builder
import org.jesperancinha.shell.webflux.data.SeaShellCostumeDto
import org.jesperancinha.shell.webflux.repo.ShellLowerRepositoryImpl
import org.jesperancinha.shell.webflux.service.SeaShellConverter
import java.util.concurrent.RecursiveTask

@Builder
@AllArgsConstructor
class SeaShellLowerRecursiveTask(
    private val lowerRepository: ShellLowerRepositoryImpl,
    private val costumeDto: SeaShellCostumeDto
) : RecursiveTask<SeaShellCostumeDto>() {

    override fun compute(): SeaShellCostumeDto =
        costumeDto.copy(lowerDto = costumeDto.lowerId
            ?.let { lowerRepository.findLowerByIdBlock(it) }
            ?.let { SeaShellConverter.toLowerDto(it) })
}